### PR TITLE
stb_image: fix implicit conversion warnings

### LINKED ISSUE #1444: stb_image: warnings MSVC: Implicity conversions/possible loss of data
**Describe the bug**
After some recent changes to stb_image.h file, when compiling with MSVC set to treat all warnings as errors, this library fails to compile at these lines with this message: `warning C4244: 'argument': conversion from '...' to '...', possible loss of data`
```
- stb_image.h(2224,32)
- stb_image.h(2281,35)
- stb_image.h(3425,49)
```

**Expected behavior**
Compilation with no warnings, the easiest fix would be to explicitly cast these variables.

Linked my pr as I am trying to fix it.

### PR DESCRIPTION
**Pull request by diamante0018**
Hello, first-time contributor here.
This pull request fixes 3 warnings that I get when using this library on the MSVC platform.
I have set my MSVC compiler to treat these warnings as errors so this is necessary for me to continue using this library without making any changes to the solution file on visual studio/turning off warnings.
Thanks.

Closes #1444

### CODE REVIEW

**Comment by nothings** [comment]
None of these seem like the correct way to fix the issue. For example, for security reasons, the code wants to check that multiplying two shorts gives a valid result. If one of the inputs isn't a short, but you cast it to a short for just that function call, you could truncate it smaller than it is, the security test passed, but then the actual multiply is done with a larger number, breaking the security check.

**Comment by diamante0018** [comment]
So, if I understand this correctly, instead of casting you would prefer I clamp the value to be in between short min/max for that function call
(And unsigned char for the final warning)?
Is there a place in this repository where it's done already so I may have a look at it?

**Comment by nothings** [comment]
I'm pretty much saying that @rygorous (who wrote this code) should take a look at it and figure out what's going on. Presumably either `dc` ought to be stored in a `short`, or `stbi__mul2shorts_valid` should take ints, or something.

**Comment by diamante0018** [comment]
Roger, feel free to close this pr when it is fixed properly or I will come back to it in a bit to see when it has been fixed to close it myself 🙂

**Comment by rygorous** [comment]
I didn't write it, Neil did, but I'll have a look later.

File: stb_image.h
Comment (NeilBickford-NV) [inline stb_image.h:2224]: Oh yep, this was my fault in #1297 ! Thank you for the fix. How about checking that `dc` is valid to cast to a short, and casting the `stbi__uint16 as well, like this?
```
if ((dc > SHRT_MAX) || (dequant[0] > SHRT_MAX) || !stbi__mul2shorts_valid((short) dc, (short)dequant[0])) return stbi__err("can't merge dc and ac", "Corrupt JPEG");
```

File: stb_image.h
Comment (NeilBickford-NV) [inline stb_image.h:2281]: Similarly, maybe this could check as well that `dc` can be casted to a short before doing so, like this?
```
if ((dc > SHRT_MAX) || !stbi__mul2shorts_valid((short) dc, 1 << j->succ_low)) return stbi__err("can't merge dc and ac", "Corrupt JPEG");
```

**Comment by diamante0018** [comment]
Thanks for the feedback, let me know if the last cast to `unsigned char` is fine.
Also I'm not familiar with the code style of this project but I tried my best to follow it, please let me know if it is okay.

**Comment by NeilBickford-NV** [comment]
Thank you, that looks good to me from a security perspective - `stbi__skip_jpeg_junk_at_end()` always returns a value between 0 and 0xff inclusive, so another solution is it could return a `stbi_uc` instead, which would avoid the cast to unsigned char.

By the way, Sean and Fabian, please feel free to @ me any time - I'm always happy to help with things!

**Comment by diamante0018** [comment]
You are right! I changed the return type of stbi__skip_jpeg_junk_at_end and removed the cast entirely.
Thanks.

**Review by NeilBickford-NV — APPROVED** [review]
The changes look good to me!

Ah, I don't have permission to merge anything in, though, and am not a project member, sorry - that would be @nothings and @rygorous . I understand it usually takes some time for things to be merged in - what I've usually done in the past for projects that require a patched version of `stb_image` is I've pointed Git submodules to the fork used for the merge request (or patched the copy if the project is using a copy of the file), then pointed them back to `nothings:master` once the patch has been merged.

**Comment by diamante0018** [comment]
Thanks for the tip, I will do that I think. Or just suppress the warning before including the headers.

**Comment by ChrisThrasher** [comment]
I'd like to bump this PR since I'm dealing with the same problems in #1444. @nothings Is this good to merge?

**Comment by diamante0018** [comment]
Woot

**Comment by rygorous** [comment]
A fix for this is already in the dev branch (and has been since May 2nd, commit https://github.com/nothings/stb/commit/8c3aa0548766e6fef7da184107d35fd81315c1a5), it's just waiting on the next time we do a release.

**Comment by diamante0018** [comment]
Better late then never. May I suggest that some MSVC tests are added to STB workflow(s) to see if it compiles with all warnings enabled.
When I tried to recently compile stb (a week ago or so) it still failed.

**Pull request closed without merging**
